Source/DafnyLanguageServer.Test/Diagnostics/SimpleLinearVerificationDiagnosticTester.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Extensions;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Util;
+using Microsoft.Dafny.LanguageServer.Workspace.Notifications;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.IntegrationTest.Diagnostics;
+
+[TestClass]
+public class SimpleLinearVerificationDiagnosticTester : LinearVerificationDiagnosticTester {
+  private const int MaxTestExecutionTimeMs = 10000;
+
+  // To add a new test, just call VerifyTrace on a given program,
+  // the test will fail and give the correct output that can be use for the test
+  // Add '//Next<n>:' to edit a line multiple times
+
+  [TestMethod, Timeout(MaxTestExecutionTimeMs)]
+  public async Task EnsureVerificationDiagnosticsAreWorking() {
+    await VerifyTrace(@"
+ .  |  |  |  I  I  |  | :predicate Ok() {
```

Comment:
predicate `Ok` doesn't have any associated verification condition right? How come it gets any icons, and also why do its icons change?

Reply:
Here is the story. I build the verification tree by first going through the program, modules, and extract methods, subset types, and field declarations. For each of them either I take a previous verification tree if it was there, or I create a new verification tree. In either case, I set their "StatusCurrent" to "Obsolete", (either ` . ` if it's the first time or ` I ` if it was verified before)

On receiving all the implementations that are going to be verified, I observed that any of these trees without an implementation can be marked as "verified" immediately, that's why it does not go through the "CurrentStatus.Verifying". Now, let me answer your questions

> predicate Ok doesn't have any associated verification condition right? 

Yes that's right, so it's why it does not get the "verifying" state.

> How come it gets any icons ?

The Dafny pipeline verified that there is nothing to do and everything works well, without the need to go through the Verifer. It's different from "Nothing" that says the verifier doesn't know yet the state of that line.

> Why do its icons change?

Until we get the verifier's information that there is no implementation, we cannot know whether the function is verified or not in general. Moreover, when the functions are edited, they have 

Now, here is a possible optimization I want to do in the future but not this PR, which is to compute the call graph and if the method was not changed and is not part of another method's call graph that was modified, we can immediately ensure its state is currently verified and, besides not even send it to Boogie, have no intermediate state in the gutter.